### ISSUE TITLE
Provide `declare` keyword support for class methods

### ISSUE THREAD

**Issue by JasonHK** [microsoft/TypeScript#38008]
## Search Terms

declare, class, method

## Suggestion

I know that `declare` keyword already support class properties. I've wonder if it is possiable to bring the support to class methods as well?

## Use Cases

You override a method `a`, and there are some other methods that return the return value of `a`.

## Examples

```typescript
class BaseTag
{
    constructor(tag: string) {...}

    static fromString(tag: string): BaseTag
    {
        return new BaseTag(tag);
    }

    static fromNumber(tag: number): BaseTag
    {
        return this.fromString(tag.toString());
    }
}

class ExtendedTag extends BaseTag
{
    static fromString(tag: string): ExtendedTag
    {
        return new BaseTag(`Tag: ${ tag} `);
    }

    declare static fromNumber(tag: number): ExtendedTag;

    // // Without `declare` support
    // static fromNumber(tag: number): ExtendedTag
    // {
    //     return super.fromString(tag) as unknown as ExtendedTag;
    // }
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
What you're asking for is #5863

**Comment by JasonHK** [comment]
> What you're asking for is #5863

@RyanCavanaugh No, this is a different idea. Let me give you another example:

```typescript
class BaseClass
{
    declare private version: string;

    public toJSON(): BaseInterface
    {
        return {
            version: this.version,
        };
    }

    protected toObject(): ObjectMixin<BaseInterface>
    {
        return Object.assign(this.toJSON(), { hello: true });
    }
}

interface BaseInterface
{
    version: string;
}

class ExtendedClass extends BaseClass
{
    declare private count: number;

    public toJSON(): ExtendedInterface
    {
        return Object.assign(super.toJSON(), { count: this.count });
    }

    declare protected toObject(): ObjectMixin<ExtendedInterface>;

    // // Without `declare` support
    // protected toObject(): ObjectMixin<ExtendedInterface>
    // {
    //     return super.toObject() as unknown as ObjectMixin<ExtendedInterface>;
    // }
}

interface ExtendedInterface extends BaseInterface
{
    count: number;
}

type ObjectMixin<T> = T & { hello: boolean; };
```

[TypeScript Playground](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=47&pc=1#code/MYGwhgzhAEBCkFMDC4oCgDebo+gEwVDACcFoAHYgSwDcwAXMmhYiKgewDsAuaCe6pwDmAbjTZc5AK4AjEFWDR67AFIBlAPIA5ABQBKXvAgIAkp0bEAZmGAIJOLLifRS9KcU7RHzn81YceJQALKggAOj82LgAae2cAXzEnePEnSnZGYEY8JXYNGQArQnp9XnyirIBZKgAPKk4AHiNTcxZrWwA+OO8fV3dPcuKwyDYhTh16EPDldW19aK9oIIQQEHZeASkyeL0k3BSUtHqLdrJmsxObOx7oSIDefkFRNEOiKGgAURrGTgI8FBG0AQ3wQvxgzQB6BuBCIpAo1DojGgwHYUnMvE4UgAtjIWGI4tI5ApcrNdAZPiDfgg8Bc2ldunEnH0PNBBllhlAqGMdBApOQWGEZpoyQsMMjUejgqEwii0fRoDs9jhDk4YeA4elMtlcmySuTddU6o0vj8-rSrFcOvinAB6G3QO3QADqVEmqPlAAM1SQEB6+HzyOxiPQ4o7NcVqTrCsVSqzo1VavUGibQWbWhbOqH7TdHUyEG4Wbz+cRBXl43roJBoGiANacdgAd08VYNieNlLTl06Sod9oO4mOdNsFNN1PNpyBHfBiHH9JusslmJxeJe4noAE9+XGKvRDUmACodaAAXmg++gADJFstVutoDJ2OwQAgwJwRAqxEA)

Although it seems that the example can be replaced by a generic, what if there are multiple return types that had to replace?

I know I can use an interface with the name of the class to override the method, but this only works for public members, not protected or private ones.

More importantly, if `declare` keyword **already** supports class properties, why don't add support to class methods as well?

**Comment by JasonHK** [comment]
@RyanCavanaugh Should I write a proposal for this?

**Comment by whzx5byb** [comment]
`declare protected toObject: () => ObjectMixin<ExtendedInterface>;` should work?

**Comment by JasonHK** [comment]
> `declare protected toObject: () => ObjectMixin<ExtendedInterface>;` should work?

@whzx5byb Yes, this will work. But I hope `declare` methods would be supported as well.

Sorry for the late reply 😅

**Comment by JasonHK** [comment]
> `declare protected toObject: () => ObjectMixin<ExtendedInterface>;` should work?

@whzx5byb Well... There's a problem with this approach though, you cannot use a method to override a property, which means this will not work:

```ts
class SuperClass
{
    declare protected toObject: () => ObjectMixin<ExtendedInterface>;
}

class SubClass extends SuperClass
{
    protected toObject(): ObjectMixin<...>
    {
        ...
    }
}
```
